File: java/src/test/java/site/markhenrick/recreational/adventofcode/y2020/Day03Test.java
package site.markhenrick.recreational.adventofcode.y2020;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.MethodSource;
import site.markhenrick.recreational.common.TestUtil;

import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.params.provider.Arguments.arguments;

class Day03Test {
	private static final String SAMPLE_INPUT = """
..##.......
#...#...#..
.#....#..#.
..#.#...#.#
.#...##..#.
..#.##.....
.#.#.#....#
.#........#
#.##...#...
#...##....#
.#..#...#.#
""";
	private static final String MY_INPUT = TestUtil.getResourceAsString("AoC/input/2020/day03.txt");

	@ParameterizedTest
	@CsvSource({
		"0,  0,  .",
		"0,  1,  .",
		"1,  0,  #",
		"10, 0,  .",
		"0,  10, .",
		"10, 10, #",
		"5,  2,  #",
		"0,  11, .",
		"0,  12, .",
		"1,  11, #",
		"10, 11, .",
		"0,  21, .",
		"10, 21, #",
		"5,  13, #"
	})
	void getChar(final int x, final int y, final char expected) {
		final var day = new Day03(SAMPLE_INPUT);
		assertThat(day.getChar(new Day03.IntVector(x, y))).isEqualTo(expected);
	}

	@SuppressWarnings("unused")
	static Stream<Arguments> solvePart1() {
		return Stream.of(
			arguments(SAMPLE_INPUT, 7),
			arguments(MY_INPUT, 292)
		);
	}

	@ParameterizedTest
	@MethodSource
	void solvePart1(final String input, final int expected) {
		final var day = new Day03(input);
		assertThat(day.solvePart1()).isEqualTo(expected);
	}

	@SuppressWarnings("unused")
	static Stream<Arguments> solvePart2() {
		return Stream.of(
			arguments(SAMPLE_INPUT, 336L),
			arguments(MY_INPUT, 9354744432L)
		);
	}

	@ParameterizedTest
	@MethodSource
	void solvePart2(final String input, final long expected) {
		final var day = new Day03(input);
		assertThat(day.solvePart2()).isEqualTo(expected);
	}
}

File: kotlin/src/test/kotlin/site/markhenrick/recreational/adventofcode/y2024/Day03Test.kt
package site.markhenrick.recreational.adventofcode.y2024

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import site.markhenrick.recreational.adventofcode.y2024.Day03.solvePart1
import site.markhenrick.recreational.getChallengeInput
import java.util.stream.Stream

class Day03Test {
    @ParameterizedTest
    @MethodSource
    fun part1(input: String, expected: Int) {
        assertThat(solvePart1(input)).isEqualTo(expected)
    }

    companion object {
        private val sampleInput = "xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))"
        private val challengeInput = getChallengeInput(2024, 3)

        @JvmStatic
        fun part1(): Stream<Arguments> = Stream.of(
            Arguments.of(sampleInput, 161),
            Arguments.of(challengeInput, 178794710),
        )
    }
}